File: cat-client/src/main/java/com/dianping/cat/message/internal/DefaultMessageManager.java
/*
 * Copyright (c) 2011-2018, Meituan Dianping. All Rights Reserved.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dianping.cat.message.internal;

import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.Stack;
import java.util.concurrent.atomic.AtomicInteger;

import org.codehaus.plexus.logging.LogEnabled;
import org.codehaus.plexus.logging.Logger;
import org.codehaus.plexus.personality.plexus.lifecycle.phase.Initializable;
import org.codehaus.plexus.personality.plexus.lifecycle.phase.InitializationException;
import org.unidal.lookup.ContainerHolder;
import org.unidal.lookup.annotation.Inject;
import org.unidal.lookup.annotation.Named;

import com.dianping.cat.ApplicationSettings;
import com.dianping.cat.Cat;
import com.dianping.cat.configuration.ClientConfigManager;
import com.dianping.cat.configuration.NetworkInterfaceManager;
import com.dianping.cat.configuration.client.entity.Domain;
import com.dianping.cat.message.ForkedTransaction;
import com.dianping.cat.message.Message;
import com.dianping.cat.message.TaggedTransaction;
import com.dianping.cat.message.Transaction;
import com.dianping.cat.message.io.MessageSender;
import com.dianping.cat.message.io.TransportManager;
import com.dianping.cat.message.spi.MessageManager;
import com.dianping.cat.message.spi.MessageTree;
import com.dianping.cat.message.spi.internal.DefaultMessageTree;

@Named(type = MessageManager.class)
public class DefaultMessageManager extends ContainerHolder implements MessageManager, Initializable, LogEnabled {

	@Inject
	private ClientConfigManager m_configManager;

	@Inject
	private TransportManager m_transportManager;

	@Inject
	private MessageIdFactory m_factory;

	private ThreadLocal<Context> m_context = new ThreadLocal<Context>();

	private long m_throttleTimes;

	private Domain m_domain;

	private String m_hostName;

	private boolean m_firstMessage = true;

	private TransactionHelper m_validator = new TransactionHelper();

	private Map<String, TaggedTransaction> m_taggedTransactions;

	private AtomicInteger m_sampleCount = new AtomicInteger();

	private Logger m_logger;

	@Override
	public void add(Message message) {
		Context ctx = getContext();

		if (ctx != null) {
			ctx.add(message);
		}
	}

	@Override
	public void bind(String tag, String title) {
		TaggedTransaction t = m_taggedTransactions.get(tag);

		if (t != null) {
			MessageTree tree = getThreadLocalMessageTree();
			String messageId = tree.getMessageId();

			if (messageId == null) {
				messageId = nextMessageId();
				tree.setMessageId(messageId);
			}
			if (tree != null) {
				t.start();
				t.bind(tag, messageId, title);
			}
		}
	}

	@Override
	public void enableLogging(Logger logger) {
		m_logger = logger;
	}

	@Override
	public void end(Transaction transaction) {
		Context ctx = getContext();

		if (ctx != null && transaction.isStandalone()) {
			if (ctx.end(this, transaction)) {
				m_context.remove();
			}
		}
	}

	public void flush(MessageTree tree, boolean clearContext) {
		MessageSender sender = m_transportManager.getSender();

		if (sender != null && isMessageEnabled()) {
			sender.send(tree);

			if (clearContext) {
				reset();
			}
		} else {
			m_throttleTimes++;

			if (m_throttleTimes % 10000 == 0 || m_throttleTimes == 1) {
				m_logger.info("Cat Message is throttled! Times:" + m_throttleTimes);
			}
		}
	}

	@Override
	public ClientConfigManager getConfigManager() {
		return m_configManager;
	}

	private Context getContext() {
		if (Cat.isInitialized()) {
			Context ctx = m_context.get();

			if (ctx != null) {
				return ctx;
			} else {
				if (m_domain != null) {
					ctx = new Context(m_domain.getId(), m_hostName, m_domain.getIp());
				} else {
					ctx = new Context("Unknown", m_hostName, "");
				}

				m_context.set(ctx);
				return ctx;
			}
		}

		return null;
	}

	@Override
	public String getDomain() {
		return m_domain.getId();
	}

	public String getMetricType() {
		return "";
	}

	public void setMetricType(String metricType) {
	}

	@Override
	public Transaction getPeekTransaction() {
		Context ctx = getContext();

		if (ctx != null) {
			return ctx.peekTransaction(this);
		} else {
			return null;
		}
	}

	@Override
	public MessageTree getThreadLocalMessageTree() {
		Context ctx = m_context.get();

		if (ctx == null) {
			setup();
		}
		ctx = m_context.get();

		return ctx.m_tree;
	}

	@Override
	public boolean hasContext() {
		Context context = m_context.get();
		boolean has = context != null;

		if (has) {
			MessageTree tree = context.m_tree;

			if (tree == null) {
				return false;
			}
		}
		return has;
	}

	private boolean hitSample(double sampleRatio) {
		int count = m_sampleCount.incrementAndGet();

		return count % ((int) (1.0 / sampleRatio)) == 0;
	}

	@Override
	public void initialize() throws InitializationException {
		m_domain = m_configManager.getDomain();
		m_hostName = NetworkInterfaceManager.INSTANCE.getLocalHostName();

		if (m_domain.getIp() == null) {
			m_domain.setIp(NetworkInterfaceManager.INSTANCE.getLocalHostAddress());
		}

		// initialize domain and IP address
		try {
			m_factory.initialize(m_domain.getId());
		} catch (Exception e) {
			m_logger.error("error when create mark file", e);
		}

		// initialize the tagged transaction cache
		final int size = m_configManager.getTaggedTransactionCacheSize();

		m_taggedTransactions = new LinkedHashMap<String, TaggedTransaction>(size * 4 / 3 + 1, 0.75f, true) {
			private static final long serialVersionUID = 1L;

			@Override
			protected boolean removeEldestEntry(Entry<String, TaggedTransaction> eldest) {
				return size() >= size;
			}
		};
	}

	@Override
	public boolean isCatEnabled() {
		return m_domain != null && m_domain.isEnabled() && m_configManager.isCatEnabled();
	}

	@Override
	public boolean isMessageEnabled() {
		return m_domain != null && m_domain.isEnabled() && m_context.get() != null && m_configManager.isCatEnabled();
	}

	public boolean isTraceMode() {
		Context content = getContext();

		if (content != null) {
			return content.isTraceMode();
		} else {
			return false;
		}
	}

	public void setTraceMode(boolean traceMode) {
		Context context = getContext();

		if (context != null) {
			context.setTraceMode(traceMode);
		}
	}

	public void linkAsRunAway(DefaultForkedTransaction transaction) {
		Context ctx = getContext();
		if (ctx != null) {
			ctx.linkAsRunAway(transaction);
		}
	}

	private String nextMessageId() {
		return m_factory.getNextId();
	}

	@Override
	public void reset() {
		// destroy current thread local data
		Context ctx = m_context.get();

		if (ctx != null) {
			if (ctx.m_totalDurationInMicros == 0) {
				ctx.m_stack.clear();
				ctx.m_knownExceptions.clear();
				m_context.remove();
			} else {
				ctx.m_knownExceptions.clear();
			}

			MessageTree tree = ctx.m_tree;
			if (tree != null) {
				double samplingRate = m_configManager.getSampleRatio();

				if (samplingRate < 1.0 && hitSample(samplingRate)) {
					tree.setHitSample(true);
				} else {
					tree.setHitSample(false);
				}
			}
		}
	}

	@Override
	public void setup() {
		Context ctx;

		if (m_domain != null) {
			ctx = new Context(m_domain.getId(), m_hostName, m_domain.getIp());
		} else {
			ctx = new Context("Unknown", m_hostName, "");
		}
		double samplingRate = m_configManager.getSampleRatio();

		if (samplingRate < 1.0 && hitSample(samplingRate)) {
			ctx.m_tree.setHitSample(true);
		}
		m_context.set(ctx);
	}

	boolean shouldLog(Throwable e) {
		Context ctx = m_context.get();

		if (ctx != null) {
			return ctx.shouldLog(e);
		} else {
			return true;
		}
	}

	@Override
	public void start(Transaction transaction, boolean forked) {
		Context ctx = getContext();

		if (ctx != null) {
			ctx.start(transaction, forked);

			if (transaction instanceof TaggedTransaction) {
				TaggedTransaction tt = (TaggedTransaction) transaction;

				m_taggedTransactions.put(tt.getTag(), tt);
			}
		} else if (m_firstMessage) {
			m_firstMessage = false;
			m_logger.warn("CAT client is not enabled because it's not initialized yet");
		}
	}

	class Context {
		private MessageTree m_tree;

		private Stack<Transaction> m_stack;

		private int m_length;

		private boolean m_traceMode;

		private long m_totalDurationInMicros; // for truncate message

		private Set<Throwable> m_knownExceptions;

		public Context(String domain, String hostName, String ipAddress) {
			m_tree = new DefaultMessageTree();
			m_stack = new Stack<Transaction>();

			Thread thread = Thread.currentThread();
			String groupName = thread.getThreadGroup().getName();

			m_tree.setThreadGroupName(groupName);
			m_tree.setThreadId(String.valueOf(thread.getId()));
			m_tree.setThreadName(thread.getName());

			m_tree.setDomain(domain);
			m_tree.setHostName(hostName);
			m_tree.setIpAddress(ipAddress);
			m_length = 1;
			m_knownExceptions = new HashSet<Throwable>();
		}

		public void add(Message message) {
			if (m_stack.isEmpty()) {
				MessageTree tree = m_tree.copy();

				tree.setMessage(message);
				flush(tree, true);
			} else {
				Transaction parent = m_stack.peek();

				addTransactionChild(message, parent);
			}
		}

		private void addTransactionChild(Message message, Transaction transaction) {
			long treePeriod = trimToHour(m_tree.getMessage().getTimestamp());
			long messagePeriod = trimToHour(message.getTimestamp() - 10 * 1000L); // 10 seconds extra time allowed

			if (treePeriod < messagePeriod || m_length >= ApplicationSettings.getTreeLengthLimit()) {
				m_validator.truncateAndFlush(this, message.getTimestamp());
			}

			transaction.addChild(message);
			m_length++;
		}

		private void adjustForTruncatedTransaction(Transaction root) {
			DefaultEvent next = new DefaultEvent("TruncatedTransaction", "TotalDuration");
			long actualDurationInMicros = m_totalDurationInMicros + root.getDurationInMicros();

			next.addData(String.valueOf(actualDurationInMicros));
			next.setStatus(Message.SUCCESS);
			root.addChild(next);

			m_totalDurationInMicros = 0;
		}

		/**
			* return true means the transaction has been flushed.
			*
			* @param manager
			* @param transaction
			* @return true if message is flushed, false otherwise
			*/
		public boolean end(DefaultMessageManager manager, Transaction transaction) {
			if (!m_stack.isEmpty()) {
				Transaction current = m_stack.pop();

				if (transaction == current) {
					m_validator.validate(m_stack.isEmpty() ? null : m_stack.peek(), current);
				} else {
					while (transaction != current && !m_stack.empty()) {
						m_validator.validate(m_stack.peek(), current);

						current = m_stack.pop();
					}
				}

				if (m_stack.isEmpty()) {
					MessageTree tree = m_tree.copy();

					m_tree.setMessageId(null);
					m_tree.setMessage(null);

					if (m_totalDurationInMicros > 0) {
						adjustForTruncatedTransaction((Transaction) tree.getMessage());
					}

					manager.flush(tree, true);
					return true;
				}
			}

			return false;
		}

		public boolean isTraceMode() {
			return m_traceMode;
		}

		public void setTraceMode(boolean traceMode) {
			m_traceMode = traceMode;
		}

		public void linkAsRunAway(DefaultForkedTransaction transaction) {
			m_validator.linkAsRunAway(transaction);
		}

		public Transaction peekTransaction(DefaultMessageManager defaultMessageManager) {
			if (m_stack.isEmpty()) {
				return null;
			} else {
				return m_stack.peek();
			}
		}

		public boolean shouldLog(Throwable e) {
			if (m_knownExceptions == null) {
				m_knownExceptions = new HashSet<Throwable>();
			}

			if (m_knownExceptions.contains(e)) {
				return false;
			} else {
				m_knownExceptions.add(e);
				return true;
			}
		}

		public void start(Transaction transaction, boolean forked) {
			if (!m_stack.isEmpty()) {
				// Do NOT make strong reference from parent transaction to forked transaction.
				// Instead, we create a "soft" reference to forked transaction later, via linkAsRunAway()
				// By doing so, there is no need for synchronization between parent and child threads.
				// Both threads can complete() anytime despite the other thread.
				if (!(transaction instanceof ForkedTransaction)) {
					Transaction parent = m_stack.peek();
					addTransactionChild(transaction, parent);
				}
			} else {
				m_tree.setMessage(transaction);
			}

			if (!forked) {
				m_stack.push(transaction);
			}
		}

		private long trimToHour(long timestamp) {
			return timestamp - timestamp % (3600 * 1000L);
		}
	}

	class TransactionHelper {
		private void linkAsRunAway(DefaultForkedTransaction transaction) {
			DefaultEvent event = new DefaultEvent("RemoteCall", "RunAway");

			event.addData(transaction.getForkedMessageId(), transaction.getType() + ":" + transaction.getName());
			event.setTimestamp(transaction.getTimestamp());
			event.setStatus(Message.SUCCESS);
			event.setCompleted(true);
			transaction.setStandalone(true);

			add(event);
		}

		private void markAsNotCompleted(DefaultTransaction transaction) {
			DefaultEvent event = new DefaultEvent("cat", "BadInstrument");

			event.setStatus("TransactionNotCompleted");
			event.setCompleted(true);
			transaction.addChild(event);
			transaction.setCompleted(true);
		}

		private void markAsRunAway(Transaction parent, DefaultTaggedTransaction transaction) {
			if (!transaction.hasChildren()) {
				transaction.addData("RunAway");
			}

			transaction.setStatus(Message.SUCCESS);
			transaction.setStandalone(true);
			transaction.complete();
		}

		private void migrateMessage(Stack<Transaction> stack, Transaction source, Transaction target, int level) {
			Transaction current = level < stack.size() ? stack.get(level) : null;
			boolean shouldKeep = false;

			for (Message child : source.getChildren()) {
				if (child != current) {
					target.addChild(child);
				} else {
					DefaultTransaction cloned = new DefaultTransaction(current.getType(), current.getName(),
											DefaultMessageManager.this);

					cloned.setTimestamp(current.getTimestamp());
					cloned.setDurationInMicros(current.getDurationInMicros());
					cloned.addData(current.getData().toString());
					cloned.setStatus(Message.SUCCESS);

					target.addChild(cloned);
					migrateMessage(stack, current, cloned, level + 1);
					shouldKeep = true;
				}
			}

			source.getChildren().clear();

			if (shouldKeep) { // add it back
				source.addChild(current);
			}
		}

		public void truncateAndFlush(Context ctx, long timestamp) {
			MessageTree tree = ctx.m_tree;
			Stack<Transaction> stack = ctx.m_stack;
			Message message = tree.getMessage();

			if (message instanceof DefaultTransaction) {
				String id = tree.getMessageId();

				if (id == null) {
					id = nextMessageId();
					tree.setMessageId(id);
				}

				String rootId = tree.getRootMessageId();
				String childId = nextMessageId();
				DefaultTransaction source = (DefaultTransaction) message;
				DefaultTransaction target = new DefaultTransaction(source.getType(), source.getName(), DefaultMessageManager.this);

				target.setTimestamp(source.getTimestamp());
				target.setDurationInMicros(source.getDurationInMicros());
				target.addData(source.getData().toString());
				target.setStatus(Message.SUCCESS);

				migrateMessage(stack, source, target, 1);

				for (int i = stack.size() - 1; i >= 0; i--) {
					DefaultTransaction t = (DefaultTransaction) stack.get(i);

					t.setTimestamp(timestamp);
					t.setDurationStart(System.nanoTime());
				}

				DefaultEvent next = new DefaultEvent("RemoteCall", "Next");

				next.addData(childId);
				next.setStatus(Message.SUCCESS);
				target.addChild(next);

				// tree is the parent, and m_tree is the child.
				MessageTree t = tree.copy();

				t.setMessage(target);

				ctx.m_tree.setMessageId(childId);
				ctx.m_tree.setParentMessageId(id);
				ctx.m_tree.setRootMessageId(rootId != null ? rootId : id);

				ctx.m_length = stack.size();
				ctx.m_totalDurationInMicros = ctx.m_totalDurationInMicros + target.getDurationInMicros();

				flush(t, false);
			}
		}

		public void validate(Transaction parent, Transaction transaction) {
			if (transaction.isStandalone()) {
				List<Message> children = transaction.getChildren();
				int len = children.size();

				for (int i = 0; i < len; i++) {
					Message message = children.get(i);

					if (message instanceof Transaction) {
						validate(transaction, (Transaction) message);
					}
				}

				if (!transaction.isCompleted() && transaction instanceof DefaultTransaction) {
					// missing transaction end, log a BadInstrument event so that
					// developer can fix the code
					markAsNotCompleted((DefaultTransaction) transaction);
				}
			} else if (!transaction.isCompleted()) {
				if (transaction instanceof DefaultForkedTransaction) {
					// link it as run away message since the forked transaction is not completed yet
					linkAsRunAway((DefaultForkedTransaction) transaction);
				} else if (transaction instanceof DefaultTaggedTransaction) {
					// link it as run away message since the forked transaction is not completed yet
					markAsRunAway(parent, (DefaultTaggedTransaction) transaction);
				}
			}
		}
	}
}

File: lib/csharp/src/CatClient/Message/Spi/Internals/DefaultMessageManager.cs
using System.Globalization;
using Org.Unidal.Cat.Message.Internals;
using Org.Unidal.Cat.Message.Spi.IO;
using Org.Unidal.Cat.Util;
using Org.Unidal.Cat.Configuration;
using System;
using System.Collections.Generic;
using System.Threading;
using System.Linq;

namespace Org.Unidal.Cat.Message.Spi.Internals
{
    public class DefaultMessageManager : IMessageManager
    {
        // we don't use static modifier since MessageManager is a singleton in
        // production actually
        private readonly CatThreadLocal<Context> _mContext = new CatThreadLocal<Context>();

        private AbstractClientConfig _mClientConfig;

        private MessageIdFactory _mFactory;

        private bool _mFirstMessage = true;

        private TransactionHelper _mValidator;

        private IMessageSender _mSender;

        private IMessageStatistics _mStatistics;

        private StatusUpdateTask _mStatusUpdateTask;

        private LRUCache<String, ITaggedTransaction> _mTaggedTransactions;

        // TODO implement enableLogging(Logger logger)

        #region IMessageManager Members

        public virtual AbstractClientConfig ClientConfig
        {
            get { return _mClientConfig; }
        }

        public virtual IMessageTree ThreadLocalMessageTree
        {
            get
            {
                Context ctx = _mContext.Value;

                return ctx != null ? ctx.Tree : null;
            }
        }

        public IMessageStatistics Statistics { get {return _mStatistics;}  }

        public virtual void Reset()
        {
            // destroy current thread local data
            Context ctx = _mContext.Value;

            if (ctx != null)
            {
                // For accuracy of log view, we should not remove context with zero duration.
                if (ctx._mTotalDurationInMicros == 0)
                {
                    ctx.ClearStack();
                    _mContext.Remove();
                }
                ctx._mKnownExceptions.Clear();
            }
        }

        public DefaultMessageManager()
        {
            this._mValidator = new TransactionHelper(this);
        }

        public virtual void Initialize(AbstractClientConfig clientConfig)
        {
            _mClientConfig = clientConfig;
            _mStatistics = new DefaultMessageStatistics();

            _mFactory = new MessageIdFactory();
            // initialize domain and ip address
            _mFactory.Initialize(_mClientConfig.Domain.Id);

            _mSender = new TcpMessageSender(_mClientConfig, _mStatistics, _mFactory);
            _mSender.Initialize();

            _mStatusUpdateTask = new StatusUpdateTask(_mStatistics, _mClientConfig);

            _mTaggedTransactions = new LRUCache<string, ITaggedTransaction>(CatConstants.TAGGED_TRANSACTION_CACHE_SIZE);

            // start status update task
            Thread statusUpdateTask = new Thread(_mStatusUpdateTask.Run);
            statusUpdateTask.IsBackground = true;
            statusUpdateTask.Start();
            Logger.Info("Thread(StatusUpdateTask) started.");
        }

        public virtual bool HasContext()
        {
            return _mContext.Value != null;
        }

        public virtual bool CatEnabled
        {
            get { return _mClientConfig.Domain != null && _mClientConfig.Domain.Enabled && _mContext.Value != null; }
        }

        public virtual void Add(IMessage message)
        {
            Context ctx = GetContext();

            if (ctx != null)
            {
                ctx.Add(message);
            }
            else
                Logger.Warn("Context没取到");
        }

        public void Bind(string tag, string title)
        {
            ITaggedTransaction t = _mTaggedTransactions.Get(tag);
            if (t != null)
            {
                IMessageTree tree = ThreadLocalMessageTree;
                if (null != tree)
                {
                    t.Start();
                    t.Bind(tag, tree.MessageId, title);
                }
            }
        }

        // TODO tagged transaction binding
        // public override void bind(string tag, string title);

        public virtual void Setup()
        {
            Context ctx = new Context(this, _mClientConfig.Domain.Id);

            _mContext.Value = ctx;
        }

        public bool ShouldLog(Exception e)
        {
            Context ctx = _mContext.Value;
            if (ctx != null)
            {
                return ctx.ShouldLog(e);
            }
            else
            {
                return true;
            }
        }

        public virtual void Start(ITransaction transaction, bool forked)
        {
            // Catch all exceptions because Start could be used as public API
            try
            {
                Context ctx = GetContext();

                if (ctx != null)
                {
                    ctx.Start(transaction, forked);

                    if (transaction is ITaggedTransaction)
                    {
                        ITaggedTransaction tt = (ITaggedTransaction)transaction;
                        _mTaggedTransactions.Set(tt.Tag, tt);
                    }
                }
                else if (_mFirstMessage)
                {
                    _mFirstMessage = false;
                    Logger.Info("CAT client is not enabled because it's not initialized yet");
                }
                else
                    Logger.Warn("Failed to get current context to start a transaction with. transaction: " + transaction);
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }

        public virtual void End(ITransaction transaction)
        {
            Context ctx = GetContext();

            if (ctx != null && (transaction.Standalone || transaction is DefaultTransaction))
            {
                if (ctx.End(transaction))
                {
                    _mContext.Remove();
                }
            }
        }

        #endregion

        public string CreateMessageId() 
        {
            return _mFactory.GetNextId();
        }

        public ITransaction PeekTransaction()
        {
            try
            {
                Context ctx = GetContext();
                if (null != ctx)
                {
                    return ctx.PeekTransaction();
                }
                else
                {
                    return new NullTransaction();
                }
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
                return new NullTransaction();
            }
        }

        internal void Flush(IMessageTree tree)
        {
            if (_mSender != null)
            {
                _mSender.Send(tree);

                if (_mStatistics != null)
                {
                    _mStatistics.OnSending(tree);
                }
                Reset();
            }
        }

        public void LinkAsRunAway(DefaultForkedTransaction transaction) {
            try
            {
                Context ctx = GetContext();
                if (ctx != null)
                {
                    _mValidator.LinkAsRunAway(transaction);
                }
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }

        internal Context GetContext()
        {
            if (Cat.IsInitialized())
            {
                Context ctx = _mContext.Value;

                if (ctx != null)
                {
                    return ctx;
                }
                if (_mClientConfig.DevMode)
                {
                    throw new Exception(
                        "Cat has not been initialized successfully, please call Cal.setup(...) first for each thread.");
                }
            }

            return null;
        }

        internal String NextMessageId()
        {
            return _mFactory.GetNextId();
        }

        #region Nested type: Context

        internal class Context
        {
            private readonly IMessageTree _mTree;

            private readonly ThreadSafeStack<ITransaction> _mStack;

            public int _mLength;

            public long _mTotalDurationInMicros;

            public ISet<Exception> _mKnownExceptions;
            
            private readonly DefaultMessageManager _mManager;

            public Context(DefaultMessageManager manager, String domain)
            {
                _mManager = manager;

                _mTree = new DefaultMessageTree();
                _mStack = new ThreadSafeStack<ITransaction>();

                Thread thread = Thread.CurrentThread;
                String groupName = Thread.GetDomain().FriendlyName;

                _mTree.ThreadGroupName = groupName;
                _mTree.ThreadId = thread.ManagedThreadId.ToString();
                _mTree.ThreadName = thread.Name;

                _mTree.Domain = domain;
                // _mTree.HostName = hostName;
                // _mTree.IpAddress = ipAddress;

                _mLength = 1;

                _mKnownExceptions = new HashSet<Exception>();
            }

            public IMessageTree Tree
            {
                get { return _mTree; }
            }

            public ThreadSafeStack<ITransaction> Stack
            {
                get { return _mStack; }
            }

            /// <summary>
            ///   添加Event和Heartbeat
            /// </summary>
            /// <param name="manager"> </param>
            /// <param name="message"> </param>
            public void Add(IMessage message)
            {
                if ((_mStack.Count == 0))
                {
                    IMessageTree tree = _mTree.Copy();

                    if (String.IsNullOrWhiteSpace(tree.MessageId))
                    {
                        tree.MessageId = _mManager.NextMessageId();
                    }

                    tree.Message = message;
                    tree.EstimatedByteSize = message.EstimateByteSize();
                    _mTree.EstimatedByteSize = 0;
                    _mManager.Flush(tree);
                }
                else
                {
                    ITransaction parent = _mStack.Peek();
                    AddTransactionChild(message, parent);
                }
                Tree.EstimatedByteSize += message.EstimateByteSize();
            }

            private void AddTransactionChild(IMessage message, ITransaction transaction)
            {
                long treePeriod = trimToHour(_mTree.Message.Timestamp);
                long messagePeriod = trimToHour(message.Timestamp - 10 * 1000L);

                if (treePeriod < messagePeriod || _mLength >= _mManager._mClientConfig.Domain.MaxMessageSize)
                {
                    _mManager._mValidator.TruncateAndFlush(this, message.Timestamp);
                }

                transaction.AddChild(message);
                _mLength++;
            }

            private void AdjustForTruncatedTransaction(ITransaction root)
            {
                DefaultEvent next = new DefaultEvent("TruncatedTransaction", "TotalDuration");
                long actualDurationInMicros = _mTotalDurationInMicros + root.DurationInMicros;

                next.AddData(Convert.ToString(actualDurationInMicros));
                next.Status = CatConstants.SUCCESS;
                root.AddChild(next);

                _mTotalDurationInMicros = 0;
            }

            ///<summary>
            ///  return true means the transaction has been flushed.
            ///</summary>
            ///<param name="manager"> </param>
            ///<param name="transaction"> </param>
            ///<returns> true if message is flushed, false otherwise </returns>
            public bool End(ITransaction transaction)
            {
                lock (_mStack.SyncRoot)
                {
                    if (_mStack.Count != 0)
                    {
                        ITransaction current = _mStack.Peek();

                        if (transaction == current)
                        {
                            _mStack.Pop();
                            _mManager._mValidator.Validate(_mStack.Count == 0 ? null : _mStack.Peek(), current);
                        }
                        else
                        {
                            if (_mStack.FirstOrDefault(item => item == transaction) != null)
                            {
                                current = _mStack.Pop();
                                while (transaction != current && _mStack.Count != 0)
                                {
                                    _mManager._mValidator.Validate(_mStack.Peek(), current);

                                    current = _mStack.Pop();
                                }
                            }
                        }

                        if (_mStack.Count == 0)
                        {
                            IMessageTree tree = _mTree.Copy();
                            _mTree.MessageId = null;
                            _mTree.Message = null;
                            _mTree.EstimatedByteSize = 0;

                            if (_mTotalDurationInMicros > 0)
                            {
                                AdjustForTruncatedTransaction((ITransaction)tree.Message);
                            }

                            _mManager.Flush(tree);
                            return true;
                        }
                    }
                    return false;
                }
            }

            /// <summary>
            ///   返回stack的顶部对象
            /// </summary>
            /// <returns> </returns>
            public ITransaction PeekTransaction()
            {
                return (_mStack.Count == 0) ? null : _mStack.Peek();
            }

            public bool ShouldLog(Exception e)
            {
                if (null == _mKnownExceptions)
                {
                    _mKnownExceptions = new HashSet<Exception>();
                }

                if (_mKnownExceptions.Contains(e))
                {
                    return false;
                }
                else
                {
                    _mKnownExceptions.Add(e);
                    return true;
                }
            }

            /// <summary>
            ///   添加transaction
            /// </summary>
            /// <param name="manager"> </param>
            /// <param name="transaction"> </param>
            public void Start(ITransaction transaction, bool forked)
            {
                if (_mStack.Count != 0)
                {
                    // In the corresponding Java code, standAlone is NOT set to false here
                    // transaction.Standalone = false;

                    // Do NOT make strong reference from parent transaction to forked transaction.
                    // Instead, we create a "soft" reference to forked transaction later, via linkAsRunAway()
                    // By doing so, there is no need for synchronization between parent and child threads.
                    // Both threads can complete() anytime despite the other thread.
                    if (!(transaction is IForkedTransaction)) {
                        ITransaction parent = _mStack.Peek();
                        AddTransactionChild(transaction, parent);
                    }
                }
                else
                {
                    if (_mTree.MessageId == null)
                    {
                        _mTree.MessageId = _mManager.NextMessageId();
                    }
                    
                    _mTree.Message = transaction;
                }

                if (!forked)
                {

                    Tree.EstimatedByteSize += transaction.EstimateByteSize();
                    _mStack.Push(transaction);
                }
            }

            public void ClearStack()
            {
                if (null != _mStack && _mStack.Count > 0)
                {
                    _mStack.Clear();
                }
            }

            public override string ToString()
            {
                return "Context. thread: " + _mTree.ThreadId;
            }

            private long trimToHour(long timestamp)
            {
                return timestamp - timestamp % (3600 * 1000L);
            }
        }

        #endregion

        class TransactionHelper
        {
            private readonly DefaultMessageManager _mManager;

            public TransactionHelper(DefaultMessageManager manager)
            {
                _mManager = manager;
            }

            public void LinkAsRunAway(DefaultForkedTransaction transaction)
            {
                DefaultEvent evt = new DefaultEvent("RemoteCall", "RunAway");

                evt.AddData(transaction.ForkedMessageId, transaction.Type + ":" + transaction.Name);
                evt.Timestamp = transaction.Timestamp;
                evt.Status = CatConstants.SUCCESS;
                evt.SetCompleted(true);
                transaction.Standalone = true;

                _mManager.Add(evt);
            }

            public void MarkAsRunAway(ITransaction parent, DefaultTaggedTransaction transaction)
            {
                if (!transaction.HasChildren())
                {
                    transaction.AddData("RunAway");
                }
                transaction.Status = CatConstants.SUCCESS;
                transaction.Standalone = true;
                transaction.Complete();
            }

            public void markAsNotCompleted(DefaultTransaction transaction)
            {
                DefaultEvent evt = new DefaultEvent("cat", "BadInstrument");

                evt.Status = "TransactionNotCompleted";
                evt.SetCompleted(true);
                transaction.AddChild(evt);
                transaction.SetCompleted(true);
            }

            private void MigrateMessage(ThreadSafeStack<ITransaction> stack, ITransaction source, ITransaction target, int level)
            {
                // Note that stack.ToArray() gives an array reversed, which is the opposite of Java.
                ITransaction[] onStackTransactions = stack.ToArray();
                ITransaction current = (level < stack.Count ? onStackTransactions[stack.Count - 1 - level] : null);
                bool shouldKeep = false;

                foreach (IMessage child in source.Children)
                {
                    if (child != current)
                    {
                        target.AddChild(child);
                    }
                    else
                    {
                        DefaultTransaction cloned = new DefaultTransaction(current.Type, current.Name, _mManager);

                        cloned.Timestamp = current.Timestamp;
                        cloned.DurationInMicros = current.DurationInMicros;
                        cloned.AddData(current.Data);
                        cloned.Status = CatConstants.SUCCESS;

                        target.AddChild(cloned);
                        MigrateMessage(stack, current, cloned, level + 1);
                        shouldKeep = true;
                    }
                }

                lock (source.Children)
                {
                    source.Children.Clear();
                }

                if (shouldKeep)
                {
                    source.AddChild(current);
                }
            }

            public void TruncateAndFlush(Context ctx, long timestamp)
            {
                IMessageTree tree = ctx.Tree;
                ThreadSafeStack<ITransaction> stack = ctx.Stack;
                IMessage message = tree.Message;

                if (message is DefaultTransaction)
                {
                    string id = tree.MessageId;
                    string rootId = tree.RootMessageId;
                    string childId = _mManager.NextMessageId();
                    DefaultTransaction source = (DefaultTransaction)message;
                    DefaultTransaction target = new DefaultTransaction(source.Type, source.Name, _mManager);

                    target.Timestamp = source.Timestamp;
                    target.DurationInMicros = source.DurationInMicros;
                    target.AddData(source.Data);
                    target.Status = CatConstants.SUCCESS;

                    MigrateMessage(stack, source, target, 1);

                    int reducedByteSize = 0;
                    foreach (ITransaction transaction in stack)
                    {
                        DefaultTransaction tran = (DefaultTransaction)transaction;
                        tran.Timestamp = timestamp;
                        reducedByteSize += transaction.EstimateByteSize();
                    }

                    DefaultEvent next = new DefaultEvent("RemoteCall", "Next");
                    next.AddData(childId);
                    next.Status = CatConstants.SUCCESS;
                    target.AddChild(next);

                    IMessageTree t = tree.Copy();

                    t.Message = target;

                    ctx.Tree.MessageId = childId;
                    ctx.Tree.ParentMessageId = id;

                    ctx.Tree.RootMessageId = (rootId != null ? rootId : id);

                    ctx._mLength = stack.Count;
                    // Update estimated byte size of the truncated tree to be the total size of all on-stack transactions.
                    ctx.Tree.EstimatedByteSize = reducedByteSize;

                    ctx._mTotalDurationInMicros = ctx._mTotalDurationInMicros + target.DurationInMicros;
                    _mManager.Flush(t);
                }
            }

            //验证Transaction
            public void Validate(ITransaction parent, ITransaction transaction)
            {
                if (transaction.Standalone)
                {
                    IList<IMessage> children = transaction.Children;
                    int len = children.Count;
                    for (int i = 0; i < len; i++)
                    {
                        IMessage message = children[i];

                        var childTransaction = message as ITransaction;
                        if (childTransaction != null)
                        {
                            Validate(transaction, childTransaction);
                        }
                    }

                    if (!transaction.IsCompleted() && transaction is DefaultTransaction)
                    {
                        // missing transaction end, log a BadInstrument event so that
                        // developer can fix the code
                        this.markAsNotCompleted((DefaultTransaction)transaction);
                    }
                }
                else if (!transaction.IsCompleted())
                {
                    if (transaction is DefaultForkedTransaction)
                    {
                        this.LinkAsRunAway((DefaultForkedTransaction)transaction);
                    }
                    else if (transaction is DefaultTaggedTransaction)
                    {
                        this.MarkAsRunAway(parent, (DefaultTaggedTransaction)transaction);
                    }
                }
            }
        }
    }
}